runtime/Java/src/org/antlr/v4/runtime/tree/Trees.java

```diff
@@ -189,6 +198,107 @@ else if ( !findTokens && t instanceof ParserRuleContext ) {
 		return getDescendants(t);
 	}
 
+	/** Return a new TerminalNode with a copy of all relevant fields
+	 *  from t and with null parent.
+	 *
+	 *  @since 4.6.1
+	 */
+	public static TerminalNodeImpl shallowCopy(TerminalNode t) {
+		return new TerminalNodeImpl(t.getSymbol());
+	}
+
+	/** Create a shallow copy of node t; the copy has same type as t. Copy
+	 *  all relevant fields to make a proper clone of t except
+	 *  the children list. The only caveat is that the error nodes from t
+	 *  are in fact copied into the clone.
+	 *
+	 *  t is not altered.
+	 *
+	 *  Because this uses reflection, it might be slower than using
+	 *  "new X()" directly.  It uses {@see ctorCache} to speed things up a lot.
+	 *
+	 *  @since 4.6.1
+	 */
+	public static ParserRuleContext shallowCopy(ParserRuleContext t) {
+		try {
+			Class<? extends ParserRuleContext> cl = t.getClass();
+			Class<?> sup = cl.getSuperclass();
+			Class<?> supSup = sup.getSuperclass();
+			if ( ctorCache==null ) { // in a multithreaded environment, some thread will win to set ctorCache
+				ctorCache =	new ConcurrentHashMap<>();
+			}
+			boolean
+				isAltLabelNode = sup!=ParserRuleContext.class && // usual case
```

Comment:
⚠️ This makes potentially invalid assumptions regarding the context super class option.